### PR TITLE
Refactor Muti-Option Preferences

### PR DESCRIPTION
**Pull request by Koranir**
## Summary
Refactors multiple preferences to remove common code.

This changes how you would normally get the preference, from `Preferences::Get<preference name>()` to `Preferences::<preference name>.Get()`. This is up to change.

(I've also made the `isOn =` for highlighting enabled preferences go consistently after the `text =`)

### CODE REVIEW

**Comment by RisingLeaf** [comment]
I honestly don't see the benefit from this, I mean sure less code, but in my opinion this it is even more complex to read to a beginner in c++ (template classes are frightening) and I the new access method does not feel very intuitive to me.

That said this is only my view but to do some actual review, please put function definitions in the implementation file.

**Comment by Koranir** [comment]
> I honestly don't see the benefit from this, I mean sure less code, but in my opinion this it is even more complex to read to a beginner in c++ (template classes are frightening) and I the new access method does not feel very intuitive to me.
> 
> That said this is only my view but to do some actual review, please put function definitions in the implementation file.

It's use templates or have a `Get()`, `GetString()`, and `Toggle()` function for _every_, _single_ multi-option preference, as well as saving from the need to have a person bumble around and try to fit another preference in with the rest of them. Beginner's in C++ won't even need to interact with the templating, they just need to follow what code is already there.

You can't put template implementations in a `.cpp` file.

**Comment by RisingLeaf** [comment]
> Also, you can't put template implementations in a .cpp file.

thats right, I forgot, we still dont implement them right inside the class definition look at Dialog.h for example

File: source/Preferences.h [hunk 1]
```diff
@@ -24,77 +24,87 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
-		NEUTRAL
 	};
 
+	enum class AlertIndicator : int_fast8_t {
+		NONE = 0,
+		AUDIO,
+		VISUAL,
+		BOTH
+	};
 	enum class AutoAim : int_fast8_t {
```
Comment (Amazinite) [inline source/Preferences.h:51]: Some of these enums have spaces between them and some don't. They all should.

File: source/Preferences.h [hunk 2]
```diff
@@ -110,11 +120,6 @@ class Preferences {
 	static void ToggleAmmoUsage();
 	static std::string AmmoUsage();
 
-	// Date format preferences.
-	static void ToggleDateFormat();
-	static DateFormat GetDateFormat();
-	static const std::string &DateFormatSetting();
```
Comment (Amazinite) [inline source/Preferences.h:116]: Why are some of these preference specific functions gone and others aren't? ToggleAmmoUsage I imagine might have something to do with the toggle ammo usage key, but what about the others? Could these not all just be changed to use static variables?

File: source/Preferences.cpp [hunk 3]
```diff
@@ -118,30 +117,29 @@ namespace {
...
+Preferences::MultiPreference<Preferences::BackgroundParallax, 2>Preferences::backgroundParallax(
+	{"off", "fancy", "fast"});
+Preferences::MultiPreference<Preferences::BoardingPriority, 0>Preferences::boardingPriority(
+	{"proximity", "value", "mixed"});
+Preferences::MultiPreference<Preferences::DateFormat, 0>Preferences::dateFormat(
+	{"dd/mm/yyyy", "mm/dd/yyyy", "yyyy-mm-dd"});
+Preferences::MultiPreference<Preferences::ExtendedJumpEffects, 0>Preferences::extendedJumpEffects(
+	{"off", "medium", "heavy"});
+Preferences::MultiPreference<Preferences::FlotsamCollection, 1>Preferences::flotsamCollection(
+	{"off", "on", "flagship only", "escorts only"});
```
Comment (Amazinite) [inline source/Preferences.cpp:142]: Feeling a little claustrophobic over here.
```suggestion
Preferences::MultiPreference<Preferences::AlertIndicator, 0> Preferences::alertIndicator(
	{"off", "audio", "visual", "both"});
Preferences::MultiPreference<Preferences::AutoAim, 2> Preferences::autoAim(
	{"off", "always on", "when firing"});
Preferences::MultiPreference<Preferences::AutoFire, 0> Preferences::autoFire(
	{"off", "on", "guns only", "turrets only"});
Preferences::MultiPreference<Preferences::BackgroundParallax, 2> Preferences::backgroundParallax(
	{"off", "fancy", "fast"});
Preferences::MultiPreference<Preferences::BoardingPriority, 0> Preferences::boardingPriority(
	{"proximity", "value", "mixed"});
Preferences::MultiPreference<Preferences::DateFormat, 0> Preferences::dateFormat(
	{"dd/mm/yyyy", "mm/dd/yyyy", "yyyy-mm-dd"});
Preferences::MultiPreference<Preferences::ExtendedJumpEffects, 0> Preferences::extendedJumpEffects(
	{"off", "medium", "heavy"});
Preferences::MultiPreference<Preferences::FlotsamCollection, 1> Preferences::flotsamCollection(
	{"off", "on", "flagship only", "escorts only"});
```

**Review by Amazinite — COMMENTED** [review]
I'm not sure how I feel about the public static variables. I don't recall us doing that anywhere else. What about accessing multi preferences like we access single preferences, through some function like `Preferences::Has("preference name")`. Perhaps `static const MultiPreference<T, int_fast8_t> &GetMulti(const std::string &name)`. (I will admit I'm not sure how generic return types would work when the exact return type depends on the input. Don't believe I've done that in C++ before.)

File: source/Preferences.h (hunk 2 above)
Comment (Koranir) [reply source/Preferences.h:116]: Those has specific logic that can't be captured as part of the template class - for example, `ToggleVsync()` skips over modes if they are not supported.

**Comment by Koranir** [comment]
> I'm not sure how I feel about the public static variables. I don't recall us doing that anywhere else. What about accessing multi preferences like we access single preferences, through some function like `Preferences::Has("preference name")`. Perhaps `static const MultiPreference<T, int_fast8_t> &GetMulti(const std::string &name)`. (I will admit I'm not sure how generic return types would work when the exact return type depends on the input. Don't believe I've done that in C++ before.)

I don't see how using public statics are any different than just calling `Get()` on something, it's the best solution that I have on hand and I'm open to any working alternatives that wouldn't cause a whole lotta code reuse.

It's impossible in C++ to have generic return types based on pseudo-runtime inputs.

You can see in previous commits that I used a `Preferences::GetMultis().<name>.Get()`, but just having the preferences be a public static has the same semantics and is easier to use.

File: source/Preferences.h [hunk 4]
```diff
@@ -139,46 +152,78 @@ class Preferences {
...
+	static bool DoAlertHelper(AlertIndicator toDo);
 
-	// Auto fire setting, either "off", "on", "guns only", or "turrets only".
-	static void ToggleAutoFire();
-	static AutoFire GetAutoFire();
-	static const std::string &AutoFireSetting();
+	static int GetPreviousSaveCount();
 
+public:
+	// Red alert siren and symbol
```
Comment (Amazinite) [inline source/Preferences.h:163]:
```suggestion
	// Red alert siren and symbol.
```

File: source/Preferences.h [hunk 5]
```diff
@@ -139,46 +152,78 @@ class Preferences {
 	static OverlayState StatusOverlaysState(OverlayType type);
 	static const std::string &StatusOverlaysSetting(OverlayType type);
 
-	// Auto aim setting, either "off", "always on", or "when firing".
-	static void ToggleAutoAim();
-	static AutoAim GetAutoAim();
-	static const std::string &AutoAimSetting();
+	// Red alert siren and symbol
```
Comment (Amazinite) [inline source/Preferences.h:155]:
```suggestion
	// Red alert siren and symbol.
```

File: source/Preferences.h [hunk 6]
```diff
@@ -139,46 +152,78 @@ class Preferences {
...
+	static bool PlayAudioAlert();
+	static bool DisplayVisualAlert();
+	static bool DoAlertHelper(AlertIndicator toDo);
 
-	// Auto fire setting, either "off", "on", "guns only", or "turrets only".
-	static void ToggleAutoFire();
-	static AutoFire GetAutoFire();
-	static const std::string &AutoFireSetting();
+	static int GetPreviousSaveCount();
```
Comment (Amazinite) [inline source/Preferences.h:161]:
```suggestion


```

File: source/Preferences.h [hunk 7]
```diff
@@ -24,32 +24,29 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
+		const T Get() const;
+		const std::string &GetString() const;
+		const int_fast8_t Index() const;
+
+		void Toggle();
+		void Load(int from);
+
+	private:
+		int_fast8_t index;
+		const std::vector<std::string> names;
```
Comment (Amazinite) [inline source/Preferences.h:42]:
```suggestion
```

File: source/Preferences.h [hunk 8]
```diff
@@ -139,46 +152,78 @@ class Preferences {
...
+	static MultiPreference<BoardingPriority, 0> boardingPriority;
+	// Date format preferences.
+	static MultiPreference<DateFormat, 0> dateFormat;
 	// Extended jump effects setting, either "off", "medium", or "heavy".
-	static void ToggleExtendedJumpEffects();
-	static ExtendedJumpEffects GetExtendedJumpEffects();
-	static const std::string &ExtendedJumpEffectsSetting();
+	static MultiPreference<ExtendedJumpEffects, 0> extendedJumpEffects;
+	// Flotsam setting, either "off", "on", "flagship only", or "escorts only".
+	static MultiPreference<FlotsamCollection, 1> flotsamCollection;
```
Comment (Amazinite) [inline source/Preferences.h:179]:
```suggestion
```

File: source/Preferences.cpp [hunk 9]
```diff
@@ -107,7 +107,6 @@ namespace {
 	private:
 		Preferences::OverlayState state = Preferences::OverlayState::OFF;
 	};
-
 	const vector<string> OverlaySetting::OVERLAY_SETTINGS = {"off", "always on", "damaged", "--", "on hit"};
```
Comment (Amazinite) [inline source/Preferences.cpp:110]:
```suggestion

	const vector<string> OverlaySetting::OVERLAY_SETTINGS = {"off", "always on", "damaged", "--", "on hit"};
```

File: source/Preferences.cpp [hunk 10]
```diff
@@ -118,30 +117,29 @@ namespace {
...
 
-	const vector<string> AUTO_AIM_SETTINGS = {"off", "always on", "when firing"};
-	int autoAimIndex = 2;
-
-	const vector<string> AUTO_FIRE_SETTINGS = {"off", "on", "guns only", "turrets only"};
-	int autoFireIndex = 0;
-
-	const vector<string> BOARDING_SETTINGS = {"proximity", "value", "mixed"};
-	int boardingIndex = 0;
+	int previousSaveCount = 3;
```
Comment (Amazinite) [inline source/Preferences.cpp:121]:
```suggestion
```

File: source/Preferences.cpp [hunk 11]
```diff
@@ -118,30 +117,29 @@ namespace {
...
 
-	const vector<string> FLOTSAM_SETTINGS = {"off", "on", "flagship only", "escorts only"};
-	int flotsamIndex = 1;
+}
```
Comment (Amazinite) [inline source/Preferences.cpp:123]:
```suggestion
```

**Review by Amazinite — APPROVED** [review]

**Review by OcelotWalrus — APPROVED** [review]

**Comment by warp-core** [comment]
Not a fan of the public statics. Not only does it violate principles like abstraction and encapsulation, but it also moves variables that hold the actual current values of some preferences out of the Preferences.cpp anonymous namespace into Preferences.h, splitting up where the preference data is stored. The new nested class and all its methods being public is also undesirable. No one outside Preferences.cpp needs to do anything but get the current value (as the enum type), get the string representation for the current value, and call the toggle method. They don't need access to the other member methods.
The default index being a template argument is also very strange. Not only is the state of these preferences stored somewhere different to all the other preferences, but now the default state is defined in the header as well, and also needs to be repeated in the .cpp file. The comments that just give the contents of the string vectors passed to the constructors on initialisation are also not great. Those comments would now need to be updated whenever those vectors are.

I would like for the variables that store the preference values to remain in the Preferences.cpp anonymous namespace, alongside all the other preferences. The class definition can also go there. The header file should probably just contain get methods for each of these preferences. This might involve having more lines, but a get method isn't very complicated and it's not like we skip out on `TypeOfX GetX() { return x; }` methods anywhere else.

https://github.com/endless-sky/endless-sky/compare/master...warp-core:Multi-option-preference-refactor-refactor?expand=1
This is an example of something closer to what I think we should do. The use of strings to resolve the "GetSpecialString" and "ToggleSpecial" methods isn't great and should be changed, whether that means having separate "GetXString" and "ToggleX" methods that just call the relevant method in the relevant MultiPreference object or something different.

**Comment by petervdmeer** [comment]
Closing this PR; the PR has been inactive for 3 months, while there is clear, actionable, consistent feedback from three senior ES members towards the author of this PR on what needs to change to get this PR merged.
@Koranir Please feel welcome to open another PR (on the preferences here, it can use the improvement, or on other parts of the code), but if you do, then please also make sure to keep the PR alive.

**Pull request closed without merging**